### PR TITLE
Feature smart metrics api

### LINKED ISSUE #1103: Backend for SMART metrics with API server
**Is your feature request related to a problem? Please describe.**
Purpose of this feature is to expose Smart metrics API in node container

**Describe the solution you'd like**
/smart and /smart/<serialNumber> endpoints shall be available inside a cluster

**Describe alternatives you've considered**
Drive CR with smart info.

**Additional context**
NA

### PR DESCRIPTION
**Pull request by Andrzej-Zukowski**
### Resolves #1103

_Describe your changes_

## PR checklist
- [ ] Add link to the issue
- [ ] Choose Project
- [ ] Choose PR label
- [ ] New unit tests added
- [ ] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing
_Provide test details_

### CODE REVIEW

File: pkg/node/smart.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
+/*
+Copyright © 2040 Dell Inc. or its subsidiaries. All Rights Reserved.
```
Comment (rafalw82) [inline pkg/node/smart.go:2]: 2040 -> 2024  ?

File: pkg/node/smart.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+)
+
+// SmartService represents smart API handler
+type SmartService struct {
+	mux    sync.Mutex
+	client api.DriveServiceClient
+	log    *logrus.Entry
+}
+
+// NewSmartService is the constructor for CmartService struct
```
Comment (rafalw82) [inline pkg/node/smart.go:36]: typo -> CmartService

File: pkg/node/smart.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+// Retrieve all discovered disks information/metrics.
+//
+// GET /smart
+func (s *SmartService) GetAllDrivesSmartInfo(ctx context.Context) (smart.GetAllDrivesSmartInfoRes, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	smartInfoResponse, err := s.client.GetAllDrivesSmartInfo(context.Background(), &api.Empty{})
+	if err != nil {
+		s.log.Errorf("Drivemgr response failure: %v", err)
```
Comment (rafalw82) [inline pkg/node/smart.go:58]: Seems like 404 is not best error code for this case. Might be misleading for end user.

File: pkg/node/smart.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+// Retrieve the disk information/metrics with the matching serial number.
+//
+// GET /smart/{serialNumber}
+func (s *SmartService) GetSmartInfo(ctx context.Context, params smart.GetSmartInfoParams) (smart.GetSmartInfoRes, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	smartInfoResponse, err := s.client.GetSmartInfo(context.Background(), &api.SmartInfoRequest{SerialNumber: params.SerialNumber})
+	if err != nil {
+		s.log.Errorf("Drivemgr response failure: %v", err)
```
Comment (rafalw82) [inline pkg/node/smart.go:81]: 404 seems not best for all the cases, can we somehow differentiate between not found drive for serial number and real error code?

File: pkg/drivemgr/idracmgr/idracmgr.go [hunk 5]
```diff
@@ -135,6 +135,17 @@ func (mgr *IDRACManager) LocateNode(action int32) error {
 	return status.Error(codes.Unimplemented, "method Locate not implemented in IDRACManager")
 }
 
+// GetSmartInfo implements GetSmartInfo method of DriveManager interface
+func (mgr *IDRACManager) GetSmartInfo(serialNumber string) (string, error) {
+	return "", status.Error(codes.Unimplemented, "method GetSmartInfo not implemented in BaseManager")
```
Comment (katarzyna-kulpa) [inline pkg/drivemgr/idracmgr/idracmgr.go:140]: please change error msg to point to idracmgr not basemgr

File: pkg/drivemgr/idracmgr/idracmgr.go [hunk 6]
```diff
@@ -135,6 +135,17 @@ func (mgr *IDRACManager) LocateNode(action int32) error {
...
+}
+
+// GetAllDrivesSmartInfo implements GetAllDrivesSmartInfo method of DriveManager interface
+func (mgr *IDRACManager) GetAllDrivesSmartInfo() (string, error) {
+	// not implemented
+	return "", status.Error(codes.Unimplemented, "method GetAllDrivesSmartInfo not implemented in BaseManager")
```
Comment (katarzyna-kulpa) [inline pkg/drivemgr/idracmgr/idracmgr.go:146]: please change error msg to point to idracmgr not basemgr

File: pkg/drivemgr/loopbackmgr/loopbackmgr.go [hunk 7]
```diff
@@ -539,6 +539,17 @@ func (mgr *LoopBackManager) LocateNode(action int32) error {
 	return nil
 }
 
+// GetSmartInfo implements GetSmartInfo method of DriveManager interface
+func (mgr *LoopBackManager) GetSmartInfo(serialNumber string) (string, error) {
+	return "", status.Error(codes.Unimplemented, "method GetSmartInfo not implemented in BaseManager")
```
Comment (katarzyna-kulpa) [inline pkg/drivemgr/loopbackmgr/loopbackmgr.go:544]: change error msg to point to loopbackmgr

File: pkg/drivemgr/loopbackmgr/loopbackmgr.go [hunk 8]
```diff
@@ -539,6 +539,17 @@ func (mgr *LoopBackManager) LocateNode(action int32) error {
...
+}
+
+// GetAllDrivesSmartInfo implements GetAllDrivesSmartInfo method of DriveManager interface
+func (mgr *LoopBackManager) GetAllDrivesSmartInfo() (string, error) {
+	// not implemented
+	return "", status.Error(codes.Unimplemented, "method GetAllDrivesSmartInfo not implemented in BaseManager")
```
Comment (katarzyna-kulpa) [inline pkg/drivemgr/loopbackmgr/loopbackmgr.go:550]: change error msg to point to loopbackmgr

**Review by katarzyna-kulpa — CHANGES_REQUESTED** [review]

File: pkg/node/smart.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+	return s
+}
+
+// GetAllDrivesSmartInfo implements get-all-drives-smart-info operation.
+//
+// Retrieve all discovered disks information/metrics.
+//
+// GET /smart
+func (s *SmartService) GetAllDrivesSmartInfo(ctx context.Context) (smart.GetAllDrivesSmartInfoRes, error) {
+	s.mux.Lock()
```
[the remark this answers is not in the document]
Comment (Andrzej-Zukowski) [reply pkg/node/smart.go:53]: Removed

File: cmd/node/main.go [hunk 10]
```diff
@@ -221,6 +212,47 @@ func main() {
...
+	if enableMetrics || enableSmart {
+		if enableMetrics {
+			grpc_prometheus.Register(csiUDSServer.GRPCServer)
+			grpc_prometheus.EnableHandlingTimeHistogram()
+			grpc_prometheus.EnableClientHandlingTimeHistogram()
+			prometheus.MustRegister(metrics.BuildInfo)
+			http.Handle(metricsPath, promhttp.Handler())
+		}
+		if enableSmart {
+			server, _ := smart.NewServer(node.NewSmartService(clientToDriveMgr, logger))
```
[the remark this answers is not in the document]
Comment (Andrzej-Zukowski) [reply cmd/node/main.go:240]: warn log added for case when multiple handlers want to register same query path

File: pkg/node/smart.go (hunk 1 above)
Comment (Andrzej-Zukowski) [reply pkg/node/smart.go:2]: done

File: pkg/node/smart.go (hunk 2 above)
Comment (Andrzej-Zukowski) [reply pkg/node/smart.go:36]: done

File: pkg/node/smart.go (hunk 3 above)
Comment (Andrzej-Zukowski) [reply pkg/node/smart.go:58]: error handling added for 400,404,500 scenarions

File: pkg/node/smart.go (hunk 4 above)
Comment (Andrzej-Zukowski) [reply pkg/node/smart.go:81]: error handling added for 400,404,500 scenarions

File: pkg/drivemgr/drivemgr.go [hunk 11]
```diff
@@ -28,4 +28,8 @@ type DriveManager interface {
 	Locate(serialNumber string, action int32) (currentStatus int32, err error)
 	// LocateNode manipulates of node's led state, which should be synced with drive's led
 	LocateNode(action int32) error
+	// GetSmartInfo gets smart info for specific drive with given serialNumber
```
[the remark this answers is not in the document]
Comment (Andrzej-Zukowski) [reply pkg/drivemgr/drivemgr.go:31]: API adjusted

File: pkg/drivemgr/loopbackmgr/loopbackmgr.go (hunk 7 above)
Comment (Andrzej-Zukowski) [reply pkg/drivemgr/loopbackmgr/loopbackmgr.go:544]: done

File: pkg/drivemgr/loopbackmgr/loopbackmgr.go (hunk 8 above)
Comment (Andrzej-Zukowski) [reply pkg/drivemgr/loopbackmgr/loopbackmgr.go:550]: done

**Review by rafalw82 — APPROVED** [review]

**Review by katarzyna-kulpa — APPROVED** [review]

**Pull request merged**